runtime/lua/vim/treesitter/_query_linter.lua

```diff
@@ -0,0 +1,182 @@
+        for _, match, _ in query:iter_matches(tree:root(), buf, 0, -1) do
+          local predicate
+          local predicate_type
+          for id, node, _ in pairs(match) do
+            local cap_id = query.captures[id]
+
+            -- perform language-independent checks only for first lang
+            if cap_id == 'error' and i == 1 then
+              add_lint_for_node(diagnostics, node, buf, 'Syntax error')
+            end
+
+            -- other checks rely on Neovim parser introspection
+            if lang and parser_info then
+              if cap_id == 'toplevel' then
+                local query_text = vim.treesitter.get_node_text(node, buf)
+                if not parse_cache[lang] then
+                  parse_cache[lang] = {}
+                end
+                local lang_cache = parse_cache[lang]
+
+                local cache_entry = lang_cache[query_text]
+                if type(cache_entry) == 'string' then
+                  add_lint_for_node(diagnostics, node, buf, cache_entry, lang)
+                elseif cache_entry == nil then
+                  local err
+                  ok, err = pcall(vim.treesitter.query.parse, lang, query_text)
+                  if not ok then
+                    add_lint_for_node(diagnostics, node, buf, err, lang)
+                  end
+                  lang_cache[query_text] = ok or err
+                end
+              elseif cap_id == 'field' then
+                local field_name = vim.treesitter.get_node_text(node, buf)
+                if not vim.tbl_contains(parser_info.fields, field_name) then
+                  add_lint_for_node(diagnostics, node, buf, 'Invalid field', lang)
+                end
+              elseif cap_id == 'node.named' or cap_id == 'node.anonymous' then
+                local node_type = vim.treesitter.get_node_text(node, buf)
+                local is_named = cap_id == 'node.named'
+
+                if not is_named then
+                  node_type = node_type:gsub('"(.*)".*$', '%1'):gsub('\\(.)', '%1')
+                end
+
+                local found = vim.tbl_contains(BUILT_IN_NODE_NAMES, node_type)
+                  or vim.tbl_contains(parser_info.symbols, function(s)
+                    return vim.deep_equal(s, { node_type, is_named })
+                  end, { predicate = true })
+
+                if not found then
+                  add_lint_for_node(diagnostics, node, buf, 'Invalid node type', lang)
+                end
+              elseif cap_id == 'predicate.name' then
+                predicate = node
+              elseif cap_id == 'predicate.type' then
+                predicate_type = node
+              end
+            end
+          end
+
+          if predicate and predicate_type then
+            local predicate_name = vim.treesitter.get_node_text(predicate, buf)
+            local type_string = vim.treesitter.get_node_text(predicate_type, buf)
+            -- Quirck of the query grammar that directives are also predicate!
```

Comment:
```suggestion
            -- Quirk of the query grammar that directives are also predicate!
```
(maybe a todo, since they shouldn't be?)